server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,186 @@
+	if err != nil {
+		return errors.Wrapf(err, "can't download file %s/%s", bucket, key)
+	}
+
+	mappings, err := MappingsFromJSON(resp)
+	if err != nil {
+		return errors.Wrap(err, "can't deserialize mappings file")
+	}
+
+	listedApps := adm.store.ListApps()
+
+	listedAppsMap := map[api.AppID]*api.App{}
+	updatedAppVersionsMap := map[api.AppID]string{}
+	// Update apps
+	for _, listedApp := range listedApps {
+		listedAppsMap[listedApp.Manifest.AppID] = listedApp
+		manifest, ok := mappings.Apps[listedApp.Manifest.AppID]
+		if !ok {
+			// TODO should deleting the app be that easy?
+			if err := adm.DeleteApp(listedApp); err != nil {
+				return errors.Wrapf(err, "can't delete an app")
+			}
+		}
+		if manifest.Version != listedApp.Manifest.Version {
+			// update app in proxy plugin
+			oldVersion := listedApp.Manifest.Version
+			listedApp.Manifest.Version = manifest.Version
+			if err := adm.store.StoreApp(listedApp); err != nil {
+				return errors.Wrap(err, "can't store app")
+			}
+			updatedAppVersionsMap[listedApp.Manifest.AppID] = oldVersion
+		}
+	}
+
+	// Add new apps as listed
+	for appID, manifest := range mappings.Apps {
+		if _, ok := listedAppsMap[appID]; !ok {
+			if err := adm.AddApp(manifest); err != nil {
+				return errors.Wrap(err, "can't add new app as listed")
+			}
+		}
+	}
+
+	listedAppsUpgraded := adm.store.ListApps()
+
+	// call onInstanceStartup. App migration happens here
+	for _, listedApp := range listedAppsUpgraded {
+		if listedApp.Status == api.AppStatusEnabled {
+			values := map[string]string{}
+			if _, ok := updatedAppVersionsMap[listedApp.Manifest.AppID]; ok {
+				values[oldVersionKey] = updatedAppVersionsMap[listedApp.Manifest.AppID]
+			}
+			// Call onStartup the function of the app
+			if err := adm.call(listedApp, listedApp.Manifest.OnStartup, values); err != nil {
+				adm.mm.Log.Error("Can't call onStartup func of the app", "app_id", listedApp.Manifest.AppID, "err", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (adm *Admin) DeleteApp(app *api.App) error {
+	// Call delete the function of the app
+	if err := adm.call(app, app.Manifest.Delete, nil); err != nil {
+		return errors.Wrap(err, "delete failed")
+	}
+
+	// delete oauth app
+	conf := adm.conf.GetConfig()
+	client := model.NewAPIv4Client(conf.MattermostSiteURL)
+	client.SetToken(string(adm.adminToken))
+
+	if app.OAuth2ClientID != "" {
+		success, response := client.DeleteOAuthApp(app.OAuth2ClientID)
+		if !success || response.StatusCode != http.StatusNoContent {
+			return errors.Wrap(response.Error, "failed to delete OAuth2 App")
+		}
+	}
+
+	// delete app from proxy plugin, not removing the data
+	if err := adm.store.DeleteApp(app); err != nil {
+		return errors.Wrap(err, "can't delete app")
+	}
+
+	adm.mm.Log.Info("Deleted the app", "app_id", app.Manifest.AppID)
+
+	return nil
+}
+
+func (adm *Admin) AddApp(manifest *api.Manifest) error {
+	newApp := &api.App{}
+	newApp.Manifest = manifest
+	newApp.Status = api.AppStatusListed
```

Comment:
1/5 suggest using Registered instead of Listed (like in "app is registered for use")